src/vecs/async_collection.py

```diff
@@ -0,0 +1,723 @@
+
+    Example usage:
+
+        async with vecs.create_async_client(DB_CONNECTION) as vx:
+            collection = await vx.get_or_create_collection(name="docs", dimension=3)
+            await collection.upsert([("id1", [1, 1, 1], {"key": "value"})])
+            # Further operations on 'collection'
+
+    Public Attributes:
+        name: The name of the vector collection.
+        dimension: The dimension of vectors in the collection.
+
+    Note: Some methods of this class can raise exceptions from the `vecs.exc` module if errors occur.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        dimension: int,
+        client: AsyncClient,
+        adapter: Optional[Adapter] = None,
+    ):
+        """
+        Initializes a new instance of the `AsyncCollection` class.
+
+        During expected use, developers initialize instances of `AsyncCollection` using the
+        `vecs.AsyncClient` with `vecs.AsyncClient.get_or_create_collection(...)` rather than directly.
+
+        Args:
+            name (str): The name of the collection.
+            dimension (int): The dimension of the vectors in the collection.
+            client (AsyncClient): The async client to use for interacting with the database.
+            adapter (Adapter, optional): The adapter to use for the collection.
+        """
+        self.client = client
+        self.name = name
+        self.dimension = dimension
+        self.table = build_table(name, client.meta, dimension)
+        self._index: Optional[str] = None
+        self.adapter = adapter or Adapter(steps=[NoOp(dimension=dimension)])
+
+        reported_dimensions = set(
+            [
+                x
+                for x in [
+                    dimension,
+                    adapter.exported_dimension if adapter else None,
+                ]
+                if x is not None
+            ]
+        )
+        if len(reported_dimensions) == 0:
+            raise ArgError("Either dimension or adapter must provide a dimension.")
+        elif len(reported_dimensions) > 1:
+            raise MismatchedDimension(
+                "Dimensions reported by `dimension` argument and adapter do not match."
+            )
+
+    def __repr__(self):
+        """
+        Returns a string representation of the `AsyncCollection` instance.
+
+        Returns:
+            str: A string representation of the `AsyncCollection` instance.
+        """
+        return f'vecs.AsyncCollection(name="{self.name}", dimension={self.dimension})'
+
+    async def __len__(self) -> int:
+        """
+        Returns the number of vectors in the collection.
+
+        Returns:
+            int: The number of vectors in the collection.
+        """
+        async with self.client.AsyncSession() as sess:
+            async with sess.begin():
+                stmt = select(func.count()).select_from(self.table)
+                result = await sess.execute(stmt)
+                return result.scalar() or 0
+
```

Comment:
We should not override dunder methods, for their meaning is special cased in the python interpreter. Having a `__len__` method might make the user call `len()` on this object, only to face a `TypeError` for it returning a `coroutine` instead of an integer. This should be a `_len` method.